### PR TITLE
Implemented `Joybus` `send_receive`. Updated `Controller`.

### LINKED ISSUE #35: `Controller` struct doesn't use new `Joybus` methods

### LINKED ISSUE #39: `Joybus` trait can implement the generic allocating `send_receive` method itself
there is no reason to make structs implement the `Joybus` `sned-receive` when it will simply be an allocation followed by a call to the `no_alloc` version.

### PR DESCRIPTION
**Pull request by rhedgeco**
resolves #39
resolves #35

### CODE REVIEW

File: software/fishstick/src/joybus/joybus.rs [hunk 1]
```diff
@@ -1,12 +1,37 @@
...
+    /// does nothing and returns Ok
+    fn send(&self, _: &Packet) -> Result<(), &'static str> {
+        // do nothing! data was sent!
+        Ok(())
+    }
+
+    /// fills the `receive_packet` with one/true values
+    fn send_receive_no_alloc(&self, _: &Packet, receive_packet: &mut Packet) -> Result<(), &'static str> {
+        for item in receive_packet.iter_mut() {
+            item.commit(true)
```
Comment (BryceBeagle) [inline software/fishstick/src/joybus/joybus.rs:33]: Where is `.commit` coming from?

File: software/fishstick/src/gamecube/controller.rs [hunk 2]
```diff
@@ -1,31 +1,35 @@
...
-
-        match receive_packet == expected_receive {
-            true => Ok(()),
-            false => Err(())
-        }
+    pub fn get_state(&self) -> &Packet {
+        &self.state
     }
 
-    pub fn exists(&self) -> Result<bool, ()> {
```
[the remark this answers is not in the document]
Comment (rhedgeco) [reply software/fishstick/src/gamecube/controller.rs:24]: I guess the removing of the exists function was a bit unintentional.
And there are two state functions:
1. `get_state` just gets the stored state of the controller
2. `process_state` polls the controller to obtain a new state

I may just remove all these updates and revert back to TODO statements so we can address this in another PR. What do you think?

File: software/fishstick/src/joybus/joybus.rs (hunk 1 above)
Comment (rhedgeco) [reply software/fishstick/src/joybus/joybus.rs:33]: .commit is the method in a `BitRef` (which is what you get using `iter_mut` on a `BitArr`) that allows you to set the value of that bit

File: software/fishstick/src/gamecube/controller.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhedgeco) [reply software/fishstick/src/gamecube/controller.rs:24]: For now I just added the exists method back to the struct

File: software/fishstick/src/joybus/joybus.rs (hunk 1 above)
Comment (BryceBeagle) [reply software/fishstick/src/joybus/joybus.rs:33]: Should we just leave the `receive_packet` untouched instead?

Comment (rhedgeco) [reply software/fishstick/src/joybus/joybus.rs:33]: yeah we could do that, I was just filling the packet with 'something', but we definitely don't have to.

File: software/fishstick/src/gamecube/controller.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (BryceBeagle) [reply software/fishstick/src/gamecube/controller.rs:24]: > I guess the removing of the exists function was a bit unintentional.
> And there are two state functions:
> 
> 1. `get_state` just gets the stored state of the controller
> 2. `process_state` polls the controller to obtain a new state
>
> I may just remove all these updates and revert back to TODO statements so we can address this in another PR. What do you think?

I don't know if we want that abstraction at this layer. This was supposed to be the lowest-level abstraction to the controller protocol (before getting to hardware). So we should just support exactly what the protocol provides and nothing more, imo

[the remark this answers is not in the document]
Comment (rhedgeco) [reply software/fishstick/src/gamecube/controller.rs:24]: Sounds good to me

[the remark this answers is not in the document]
Comment (rhedgeco) [reply software/fishstick/src/gamecube/controller.rs:24]: changes made, let me know @BryceBeagle

File: software/fishstick/src/joybus/joybus.rs (hunk 1 above)
Comment (BryceBeagle) [reply software/fishstick/src/joybus/joybus.rs:33]: We can readdress this in the future

**Review by BryceBeagle — APPROVED** [review]

File: software/fishstick/src/gamecube/controller.rs [hunk 3]
```diff
@@ -9,23 +13,18 @@ impl<J: Joybus> Controller<J> {
...
-        let send_packet = &[1, 2, 3];
-        let expected_receive = [4, 5, 6];
-
-        let receive_packet = self.joybus.send_receive(send_packet)?;
-
-        match receive_packet == expected_receive {
-            true => Ok(()),
-            false => Err(())
-        }
+    pub fn exists(&self) -> Result<bool, &'static str> {
```
Comment (BryceBeagle) [inline software/fishstick/src/gamecube/controller.rs:16]: I presume the errors will become real `Error` types instead of strings in the future?

Comment (rhedgeco) [reply software/fishstick/src/gamecube/controller.rs:16]: yeah

**Pull request merged**